### ISSUE TITLE
completion for (MS) sql variable doesn't work

### ISSUE THREAD

**Issue by piomiq** [notepad-plus-plus/notepad-plus-plus#11719]
### Description of the Issue
All variables in SQL (Microsoft SQL) code are prefixed by "@". When I enter any name of variable, which was declared on top of code, then completion doesn't work, but when I remove "@" as prefix then magically I get completion list.

### Steps to Reproduce the Issue

1. In SQL put following:
declare   @myVariable Integer
declare   @myVariable2 Integer

2. Press Enter twice
3. Enter below string:
@my

### Expected Behavior
should be shown dropdown list with matching strings

### Actual Behavior
nothing happens

**Comment by rdipardo** [comment]
As explained [here](https://github.com/ScintillaOrg/lexilla/issues/85#issuecomment-1136494353):

> The `sql` lexer is most targeted at Oracle's SQL variant [ . . .]
> 
> There are other SQL variant lexers: `mysql` for [MySQL](https://github.com/notepad-plus-plus/notepad-plus-plus/blob/ddbb75f0f790e3969f9ac506b5cb58eb4739e361/lexilla/lexers/LexMySQL.cxx#L4) and `mssql` [Microsoft SQL Server](https://github.com/notepad-plus-plus/notepad-plus-plus/blob/ddbb75f0f790e3969f9ac506b5cb58eb4739e361/lexilla/lexers/LexMSSQL.cxx#L3).
> 

Notepad++ comes with the `sql` lexer *only*. The vendor-specific ones are in the source repository, but not implemented. 
In the meantime, there's [a UDL](https://github.com/notepad-plus-plus/userDefinedLanguages/blob/master/UDLs/T-SQL-MSServer2008r2_bySergeyIvanov.xml) for MS Transact-SQL (based on SQL Server 2008).

**Comment by piomiq** [comment]
> Notepad++ comes with the sql lexer only. The vendor-specific ones are in the source repository, but not implemented.
In the meantime, there's [a UDL](https://github.com/notepad-plus-plus/userDefinedLanguages/blob/master/UDLs/T-SQL-MSServer2008r2_bySergeyIvanov.xml) for MS Transact-SQL (based on SQL Server 2008).

OK. Understand. But it does mean that completion of words for given language depends on this if lexer for given language is implemented? I'm not sure, because I found that only for name of variables prefixed by "@" are not provided completion drop down list, but for other words, also key words not met in Oracle) like RAISERROR, I get suggestions in drop down list. So I don't understand what is the problem with words prefixed by "@".

**Comment by rdipardo** [comment]
Sorry for misunderstanding the issue; it really has nothing to do with the lexer module.
Auto completion is a "native" feature of N++ (Scintilla's only job is displaying and &#x2014; [more recently][0] &#x2014; styling the popup window).

I checked and N++ 8.4.1 (portable) behaves exactly as you expect:

![npp_8 4 1_mssql_autocompl](https://user-images.githubusercontent.com/59004801/170792455-a0251def-a034-45f5-9ee1-9bfa60310749.gif)

Make sure the options at Settings > Preferences > Auto-Completion resemble the screen capture below:

<img width="631" alt="npp_8 4 1_prefs_autocompl_default" src="https://user-images.githubusercontent.com/59004801/170792476-2952e763-15b7-4c25-9d49-baf46e89d56c.png">


[0]: https://github.com/notepad-plus-plus/notepad-plus-plus/pull/11523

**Comment by piomiq** [comment]
I use recent version of Notepad++, so 8.4.1 (32bit, portable)
And unfortunately completion still doesn't work. I aligned my settings as you suggested:
![image](https://user-images.githubusercontent.com/18745888/171384114-3b4dca91-bb06-4cef-a924-e14405f75e82.png)

Even I restarted N++, but still no effect.
I tested this in 2 ways.
1-st.
- Ctrl+N
- copy and paste mentioned sql code
- Language -> SQL
- type Enter to go to new line
- enter "@"
result: no drop down list
- enter '@m'
result: no drop down list
2-nd
Only one difference was that I saved new file as sql file, and called it "test.sql"

![n++_no_completion2](https://user-images.githubusercontent.com/18745888/171419664-36bf0171-bcc9-4391-9658-395e2200c7ca.gif)

Let me know please if I missed something, did something wrong or alternatively tell me what more info I should provide.

**Comment by rdipardo** [comment]
I found a way to reproduce problem. Did you add the `@` character to the "Word character list" at "Settings > Preferences > Delimiter", like this?

![npp-special-delimiters](https://user-images.githubusercontent.com/59004801/171590963-063e149c-d1b9-4882-afdd-877a7555b171.png)

If so, clear it out and repeat your steps. The auto-completion function can only match alphanumeric characters. Notice how the `@` prefix never appeared in the completion list even when it worked for me. The normal behaviour is to ignore the prefix completely and begin searching at the first letter.

**Comment by piomiq** [comment]
Yes, I did and removing this character from "Word character list" fixes issue with completion name of variables.

I did it in particular purpose. I added this character to be able to highlight (by double click) whole variable name to copy and paste it in other place. Therefore,IMO such workaround doesn't resolve the issue.

**Comment by rdipardo** [comment]
@piomiq, @ArkadiuszMichalski

Auto-completion for Microsoft SQL can now be implemented : 01bbebf

As I [said before](https://github.com/notepad-plus-plus/notepad-plus-plus/pull/13184#issuecomment-1439126221), I will not be working on it, so feel free to go ahead.

**Comment by donho** [comment]
Scintilla (the edit component used by Notepad++) use word character list to recognize the words.
So if '@' is part of word, logically "@XXX" you have typed should be in the list.
   
To construct the word list from XXX, the following RegExpr is used: `\<XXX[^ \t\n\r.,;:"(){}=<>'+!?\[\]]+` where XXX is the word which user types to trigger auto-completion, in your case is '@'. 
`\<`  matches the start of a word. A word is a sequence of alphanumeric characters or underscores, of which '@' is not part of.
As a result the word list is empty.

Unfortunately, I have no solution for it. If anyone here has one, please let me know.
